usr/src/uts/common/fs/zfs/sys/zio_crypt.h

```diff
@@ -0,0 +1,177 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License, Version 1.0 only
+ * (the "License").  You may not use this file except in compliance
+ * with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+/*
+ * Copyright (c) 2016, Datto, Inc. All rights reserved.
+ */
+
+#ifndef	_SYS_ZIO_CRYPT_H
+#define	_SYS_ZIO_CRYPT_H
+
+#include <sys/dmu.h>
+#include <sys/refcount.h>
+#include <sys/crypto/api.h>
+#include <sys/nvpair.h>
+#include <sys/avl.h>
+#include <sys/zio.h>
+
+/* forward declarations */
+struct zbookmark_phys;
+
+/* macros defining key encryption lengths */
+#define	MAX_CRYPT_KEY_LEN 32
+#define	WRAPPING_KEY_LEN 32
+#define	WRAPPING_IV_LEN 13
+#define	WRAPPING_MAC_LEN 16
+#define	L2ARC_IV_LEN 32
+#define	L2ARC_MAC_LEN 8
+#define	ZIL_MAC_LEN 8
+
+#define	SHA_256_DIGEST_LEN 32
+#define	HMAC_SHA256_KEYLEN 32
+
+#define	L2ARC_DEFAULT_CRYPT ZIO_CRYPT_AES_256_CCM
+
+#define	ZIO_NO_ENCRYPTION_NEEDED -1
+
+/* utility macros */
+#define	BITS_TO_BYTES(x) (((x) + 7) >> 3)
+#define	BYTES_TO_BITS(x) (x << 3)
+
+/* supported commands for zfs_ioc_crypto() */
+typedef enum zfs_ioc_crypto_cmd {
+	ZFS_IOC_CRYPTO_CMD_NONE = 0,
+	ZFS_IOC_CRYPTO_LOAD_KEY,
+	ZFS_IOC_CRYPTO_UNLOAD_KEY,
+	ZFS_IOC_CRYPTO_ADD_KEY,
+	ZFS_IOC_CRYPTO_REWRAP,
+} zfs_ioc_crypto_cmd_t;
+
+typedef enum zio_crypt_type {
+	ZC_TYPE_NONE = 0,
+	ZC_TYPE_CCM,
+	ZC_TYPE_GCM
+} zio_crypt_type_t;
+
+/* table of supported crypto algorithms, modes and keylengths. */
+typedef struct zio_crypt_info {
+	/* mechanism name, needed by ICP */
+	crypto_mech_name_t ci_mechname;
+
+	/* cipher mode type (GCM, CCM) */
+	zio_crypt_type_t ci_crypt_type;
+
+	/* length of the encryption key */
+	size_t ci_keylen;
+
+	/* human-readable name of the encryption alforithm */
+	char *ci_name;
+} zio_crypt_info_t;
+
+extern zio_crypt_info_t zio_crypt_table[ZIO_CRYPT_FUNCTIONS];
+
+/* physical representation of a wrapped key in the DSL Keychain */
+typedef struct dsl_crypto_key_phys {
+	/* encryption algorithm (see zio_encrypt enum) */
+	uint64_t dk_crypt_alg;
```

Comment:
You are storing this in a ZAP object as an array of uint8_t's.  Therefore this struct must have be equivalent to an array of uint8_t's, otherwise it will not byteswap correctly.

It would be good to add a comment here explaining this.